samples/nrf9160/serial_lte_modem/src/slm_util.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <stdio.h>
+#include "slm_util.h"
+
+#define PRINTABLE_ASCII(ch) (ch > 0x1f && ch < 0x7f)
+
+/**
+ * @brief Compare string ignoring case
+ */
+bool slm_util_cmp(const char *str1, const char *str2, u16_t str2_len)
+{
+	int i;
+
+	if (strlen(str1) != str2_len) {
+		return false;
+	}
+
+	for (i = 0; i < str2_len; i++) {
+		if (tolower(*(str1 + i)) != *(str2 + i)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/**
+ * @brief Compare name of AT command ignoring case
+ */
+bool slm_util_cmd_cmp(const char *cmd, const char *slm_cmd, u8_t slm_cmd_len)
+{
+	int i;
+
+	if (strlen(cmd) < slm_cmd_len) {
+		return false;
+	}
+
+	for (i = 0; i < slm_cmd_len; i++) {
+		if (toupper(*(cmd + i)) != *(slm_cmd + i)) {
+			return false;
+		}
+	}
+#if defined(CONFIG_SLM_CR_LF_TERMINATION)
+	if (strlen(cmd) > (slm_cmd_len + 2)) {
+#else
+	if (strlen(cmd) > (slm_cmd_len + 1)) {
+#endif
+		char ch = *(cmd + i);
+		/* With parameter, SET TEST, "="; READ, "?" */
+		return ((ch == '=') || (ch == '?'));
+	}
+
+	return true;
+}
+
+/**
+ * @brief Detect hexdecimal data type
+ */
+bool slm_util_hex_check(const u8_t *data, u16_t data_len)
+{
+	for (int i = 0; i < data_len; i++) {
+		char ch = *(data + i);
+
+		if (!PRINTABLE_ASCII(ch) && ch != '\r' && ch != '\n') {
```

Comment:
What if binary data consists only from bytes falling into printable ascii range? I'm afraid such check could fail in certain cases...

Reply:
Understood. The point is if the HEX data is all printable, the there is no need of hex encoding/decoding over AT channel. This check is for data from remote as AT response. If MCU side has similar check, it could receive the hex data in ASCII format properly.